service/portability-service-lambda/portability_service.py

```diff
@@ -0,0 +1,330 @@
+    """
+
+    environment_id = str(uuid.uuid4())
+
+    environments_table().put_item(
+        Item={
+            "EnvironmentId": environment_id,
+            "Url": event["url"],
+            "Name": event["name"],
+            "Schema": event["schema"],
+            "Headers": event.get("headers", "none")
+        }
+    )
+
+    return {"environment_id": environment_id}
+
+
+def environments_get(event, context):
+    """Handle GET to /environments by returning a list of registered environments."""
+
+    response = environments_table().scan()
+
+    output = []
+    for item in response["Items"]:
+        output.append({"environment_id": item["EnvironmentId"],
+                       "name": item["Name"],
+                       "url": item["Url"],
+                       "schema": item["Schema"]})
+
+    return {"environments": output}
+
+
+def portability_tests_post(event, context):
+    """Handle POST to /portability_tests by submitting workflow requests to registered external
+    environments.
+
+    event is a dict that should have keys
+        - workflow_descriptor (string)
+            The main WDL file to be executed
+        - workflow_params (string)
+            JSON string of inputs to the main WDL
+        - workflow_dependencies (list of strings)
+            List of WDL strings for dependency WDLs
+    """
+
+    # Read all the registered environments so we can iterate over them
+    all_environments = environments_table().scan()["Items"]
+
+    wdl = event["workflow_descriptor"]
+    params = event["workflow_params"]
+
+    test_id = str(uuid.uuid4())
+
+    for environment in all_environments:
+        env_id = environment["EnvironmentId"]
+
+        try:
+            workflow_id = wes_submit(
+                env_id, wdl, params, event.get("workflow_dependencies"))
+            record_test_event(test_id, env_id, "SubmissionSucceeded", workflow_id)
+        except Exception as e:
+            record_test_event(test_id, env_id, "SubmissionFailed", str(e))
+
+    return {"test_id": test_id}
+
+
+def portability_tests_get_status(event, context):
+    """Handle GET to /portability_tests/{test_id}/status."""
+
+    # Get all the db entries for a submission to an environment for this test
+    response = tests_table().query(
+        KeyConditionExpression=Key("TestId").eq(event["test_id"])
+    )
+    items = response["Items"]
+
+    # 404 if we don't find anything
+    if not items:
+        error_dict = {
+            "errorType": "NotFound",
+            "httpStatus": "404",
+            "requestId": context.aws_request_id,
+            "message": "Test {} was not found".format(event["test_id"])
+        }
+
+        raise Exception(json.dumps(error_dict))
+
+    # This is a set of all the environments to which the test workflow
+    # descriptor was submitted.
+    submitted_environment_ids = set([
+        i["EnvironmentId"] for i in items
+        if i["EventType"] in ("SubmissionSucceeded", "SubmissionFailed")])
```

Comment:
Since the submission and job related statuses are used in several places here, you could put them into variables instead

Reply:
I switched this to a couple `enum` classes.